syntax = "proto3";

package yandex.cloud.priv.ydb.v1;

import "yandex/cloud/api/operation.proto";
import "yandex/cloud/priv/access/access.proto";
import "yandex/cloud/priv/operation/operation.proto";
import "yandex/cloud/priv/validation.proto";
import "yandex/cloud/priv/ydb/backup.proto";

option go_package = "a.yandex-team.ru/ydb/public/api/client/yc_private/ydb;ydb";

// A set of methods for managing backups;.
service BackupService {
  // Returns the specified backup.
  rpc Get (GetBackupRequest) returns (Backup);

  // Returns list of backup paths.
  rpc ListPaths(ListPathsRequest) returns (ListPathsResponse);

  // Retrieves a list of backups.
  rpc List (ListBackupsRequest) returns (ListBackupsResponse);

  // Restarts the specified backup.
  rpc Restart (RestartBackupRequest) returns (operation.Operation) {
    option (yandex.cloud.api.operation) = {
      metadata: "RestartBackupMetadata"
      response: "Backup"
    };
  }

  // Deletes the specified backup.
  rpc Delete (DeleteBackupRequest) returns (operation.Operation) {
    option (yandex.cloud.api.operation) = {
      metadata: "DeleteBackupMetadata"
      response: "google.protobuf.Empty"
    };
  }

  rpc ListAccessBindings (access.ListAccessBindingsRequest) returns (access.ListAccessBindingsResponse);

  rpc SetAccessBindings (access.SetAccessBindingsRequest) returns (operation.Operation) {
    option (yandex.cloud.api.operation) = {
      metadata: "access.SetAccessBindingsMetadata"
      response: "google.protobuf.Empty"
    };
  }

  rpc UpdateAccessBindings (access.UpdateAccessBindingsRequest) returns (operation.Operation) {
    option (yandex.cloud.api.operation) = {
      metadata: "access.UpdateAccessBindingsMetadata"
      response: "google.protobuf.Empty"
    };
  }

  rpc ListByStatus (ListBackupsByStatusRequest) returns (ListBackupsByStatusResponse);

}


message ListPathsRequest {
  // Required. ID of the YDB backup.
  string backup_id = 1 [(required) = true, (length) = "<=50"];

  // The maximum number of results per page that should be returned. If the number of available
  // results is larger than `page_size`, the service returns a `next_page_token` that can be used
  // to get the next page of results in subsequent ListPaths requests.
  // Acceptable values are 0 to 1000, inclusive. Default value: 100.
  int64 page_size = 2 [(value) = "0-1000"];

  // Page token. Set `page_token` to the `next_page_token` returned by a previous ListPaths
  // request to get the next page of results.
  string page_token = 3 [(length) = "<=100"];
}

message ListPathsResponse {
  repeated string paths = 1;
  string next_page_token = 2;
}

message GetBackupRequest {
  // Required. ID of the YDB backup.
  string backup_id = 1 [(required) = true, (length) = "<=50"];
}

message ListBackupsRequest {
  string folder_id = 1;

  // The maximum number of results per page that should be returned. If the number of available
  // results is larger than `page_size`, the service returns a `next_page_token` that can be used
  // to get the next page of results in subsequent ListBackups requests.
  // Acceptable values are 0 to 1000, inclusive. Default value: 100.
  int64 page_size = 2 [(value) = "0-1000"];

  // Page token. Set `page_token` to the `next_page_token` returned by a previous ListBackups
  // request to get the next page of results.
  string page_token = 3 [(length) = "<=100"];
}

message ListBackupsResponse {
  repeated Backup backups = 1;

  // This token allows you to get the next page of results for ListBackups requests,
  // if the number of results is larger than `page_size` specified in the request.
  // To get the next page, specify the value of `next_page_token` as a value for
  // the `page_token` parameter in the next ListBackups request. Subsequent ListBackups
  // requests will have their own `next_page_token` to continue paging through the results.
  string next_page_token = 2;
}

message DeleteBackupRequest {
  string backup_id = 1;
}

message DeleteBackupMetadata {
  string backup_id = 1;
  string database_id = 2;
}

message RestartBackupRequest {
  string backup_id = 1;
}

message ListBackupsByStatusRequest {

  // Status of backup to select.
  Backup.Status backup_status = 2;

  // The maximum number of results per page that should be returned. If the number of available
  // results is larger than `page_size`, the service returns a `next_page_token` that can be used
  // to get the next page of results in subsequent ListBackupsByStatus requests.
  // Acceptable values are 0 to 1000, inclusive. Default value: 100.
  int64 page_size = 3 [(value) = "0-1000"];

  // Page token. Set `page_token` to the `next_page_token` returned by a previous ListBackupsByStatus
  // request to get the next page of results.
  string page_token = 4 [(length) = "<=100"];
}

message ListBackupsByStatusResponse {
  repeated Backup backups = 1;

  // This token allows you to get the next page of results for ListBackupsByStatus requests,
  // if the number of results is larger than `page_size` specified in the request.
  // To get the next page, specify the value of `next_page_token` as a value for
  // the `page_token` parameter in the next ListBackupsByStatus request. Subsequent ListBackupsByStatus
  // requests will have their own `next_page_token` to continue paging through the results.
  string next_page_token = 2;
}

message RestartBackupMetadata {
  string backup_id = 1;
  string database_id = 2;
}
